### PR TITLE
feat(compiler): M0 language surface - hello-argv differential corpus

### LINKED ISSUE #296: compiler M0.2 — M0 language surface: hello-argv passes the differential corpus
## Parent and status — refreshed 2026-09-22

Part of #294. Implementation is in **open PR #303**, head `e66e9bde633347a4c9dc643a5f9afb72ea62a688`, above #300. Not merged; do not duplicate the work. This body describes the reconciled M0 contract across the full stack, including later diagnostic/parity refinements. #294 holds current integration pins; no tests were rerun for this update.

## What to build

Extend the native tracer bullet until hello-argv compiles through public `compile()` and matches the Node reference under the explicit contract below.

Supported M0 surface: literals, arithmetic/string expressions, unary signs, template literals, supported named function declarations/calls, `console.log`, argv indexing/length/slicing, supported string length and numeric `process.exit(code)`. Keep sibling behavior fixtures rather than snapshots of the IR or Go text.

This is not general JS/TS support: modules, variables/control flow, closures, objects/classes and async are later work. Functions do not gain async/generator/default/rest/optional parameter support. `process.exit()` without a numeric argument, `console.error` and `console.warn` are checker errors. Direct array inspection/logging and printf-style log formatting are unsupported; array-to-string coercion in templates/concatenation remains in scope. String `.slice()` is not the supported argv-array slice operation.

## Observable behavior contract

- **argv:** the runtime synthesizes `[canonical native executable path, compile-time absolute source-entry path, ...userArgs]`. It is Node-shaped, not identical to Node's executable identity. The corpus may compare `argv[1]` because both runs use the same source fixture path, and compares user arguments normally. `argv[0]` has a dedicated native executable-path assertion, including relative invocation; do not compare it to the Node binary path.
- **Installed-command gap:** `argv[1]` currently embeds a build-machine source path. A future installed Crust/hybrid profile must define its own stable argv identity; M0 fixtures do not settle that distribution contract.
- **Output/errors:** stdout and exit status are exact; successful stderr is exact. Supported runtime exceptions compare exact class, optional code and full message after stripping only Node's recognized diagnostic envelope. Unknown stderr fails; stacks, source snippets, paths and Node banners are not reproduced.
- **Undefined/string boundaries:** missing argv entries preserve JS `undefined`; reading their length raises `TypeError`, not a native panic. TS may label an addition string-valued even when runtime operands yield a number. Preserve supported direct log/template behavior, but reject unsafe argument/return crossings unless concatenation is guaranteed. Do not trust a non-null assertion as runtime proof.
- **Exit:** `process.exit(code)` requires a number. Supported invalid numeric codes preserve the `RangeError [ERR_OUT_OF_RANGE]` class/code/message, including received-value context.

## Acceptance criteria

- [ ] hello-argv (templates + argv + exit) passes the native differential corpus.
- [ ] Supported literals, arithmetic/string expressions, templates and named function calls have passing behavior fixtures.
- [ ] argv tests follow the explicit prefix contract, including the separate native executable-identity assertion.
- [ ] Output/runtime-error tests follow the exact-vs-normalized contract above.
- [ ] Missing argv/string-boundary cases preserve supported JS behavior or fail with compiler diagnostics, never a generated Go panic.
- [ ] Unsupported direct array logging remains rejected while supported array coercion is retained.

## Dependencies

Logical predecessor: #295. PR ancestry: #300 → #303; #304 supplies the seeded number-parity gate (#297). The current number-formatting implementation already lives in the language layer; do not deliberately regress it to naive formatting while working on this ticket.

The hybrid direction in #294 does not turn an unsupported native construct into automatic interpreter fallback.

[Frontend](https://github.com/chenxin-yan/crust/blob/340fc4e4fbf5735cfb4be3e8d7bf9737a2f63aef/packages/compiler/src/frontend.ts) · [runtime](https://github.com/chenxin-yan/crust/blob/340fc4e4fbf5735cfb4be3e8d7bf9737a2f63aef/packages/compiler/runtime/runtime.go) · [corpus and argv assertions](https://github.com/chenxin-yan/crust/blob/340fc4e4fbf5735cfb4be3e8d7bf9737a2f63aef/packages/compiler/tests/corpus.test.ts) · [string/error contracts](https://github.com/chenxin-yan/crust/blob/340fc4e4fbf5735cfb4be3e8d7bf9737a2f63aef/packages/compiler/README.md)

### PR DESCRIPTION
**Pull request by chenxin-yan**
## Summary

- Lower M0 literals, arithmetic/string expressions, templates, function declarations/calls, and process APIs into the compiler IR.
- Extend the compiler-owned ambient declarations with Node-shaped argv and numeric process.exit; no Node ambient dependency.
- Reject unsafe string-typed additions at function boundaries instead of allowing accepted programs to panic natively. Keep direct logging and template coercion of those runtime values.
- Correlate literal IR types and values, represent exit explicitly, remove duplicate checks/IndexLength, and emit readable collision-free identifiers.
- Preserve exact stdout and exit status. Runtime-error tests compare full normalized error class/code/message while excluding Node stack traces, source context, paths and version banners.
- This layer contains the number-formatting implementation. #304 adds deterministic parity fuzzing and narrows array-logging scope.

## Stack

`main` → #300 → #303 → #304 → #302 → #301

Restacked on main `975d893ebf0f7e1e36b7262992a9af3ecde54c18`. This PR's verified head is `e66e9bde633347a4c9dc643a5f9afb72ea62a688`. All 50 original commits are retained across the stack, plus eight scoped fixes; no merge commits.

## Validation

- Guarded `bun install --frozen-lockfile`, with unchanged lockfile and release-age policy.
- Compiler build, typecheck, lint and format checks pass.
- 34 package tests pass on this exact head.
- Built compiler entrypoint works from an external cwd under Bun and Node 24.
- Go-free API/diagnostic checks and explicit native-suite skip behavior verified.
- Native tests use Go 1.26.5 and Node 24.20.0; number-parity layers pass both default and alternate fuzz seeds.
- Final stack passes forced, zero-cache-hit root-equivalent build/lint/format/types/tests and Bun/Node/Deno distribution smokes. Framework tests: 1,699 pass and seven unrelated conditional skips.
- Independent correctness and standards reviews found no remaining blocking source defects. Hosted CI is separate from this local evidence.

Closes #296

### CODE REVIEW

**Comment by chenxin-yan** [comment]
[stack] Merge order: #300 -> #303 -> #304 -> #302 -> #301 (part 2/5). Stacked PRs get no framework CI until retargeted to main on parent merge; all gates were run locally on the integrated stack tip.

File: packages/compiler/src/emitter.ts [hunk 1]
```diff
@@ -1,16 +1,104 @@
...
+		.map((statement) => emitStatement(statement, 1))
+		.join("\n");
+	return `func ${declaration.name}(${parameters})${returnType} {\n${statements}\n}`;
+}
+
+function emitStatement(statement: Statement, indentation: number): string {
+	const indent = "\t".repeat(indentation);
+	switch (statement.kind) {
+		case "log":
+			return `${indent}crustRuntime.Log(${statement.values.map(emitExpression).join(", ")})`;
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/emitter.ts:37]: [pi-watch] fixed in fb401b6

File: packages/compiler/runtime/runtime.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+func String(value any) string {
+	switch value := value.(type) {
+	case string:
+		return value
+	case bool:
+		return strconv.FormatBool(value)
+	case float64:
+		return strconv.FormatFloat(value, 'g', -1, 64)
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] rejected: duplicate of ECMAScript number parity already implemented and fuzz-tested in stacked child PR 304

File: packages/compiler/src/emitter.ts [hunk 3]
```diff
@@ -1,16 +1,104 @@
...
+				return `crustRuntime.Exit(${emitExpression(expression.arguments[0]!)})`;
+			}
+			return `${expression.callee}(${expression.arguments.map(emitExpression).join(", ")})`;
+		case "argv":
+			return "crustRuntime.Argv()";
+		case "slice":
+			return `${emitExpression(expression.value)}[int(${emitExpression(expression.start)}):]`;
+		case "length":
+			return `float64(len(${emitExpression(expression.value)}))`;
+		case "index":
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/emitter.ts]: [pi-watch] fixed in fb401b6

File: packages/compiler/src/emitter.ts [hunk 4]
```diff
@@ -1,16 +1,104 @@
...
+			return `${emitExpression(expression.value)}[int(${emitExpression(expression.index)})]`;
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/emitter.ts]: [pi-watch] fixed in fb401b6

File: packages/compiler/src/emitter.ts [hunk 5]
```diff
@@ -1,16 +1,104 @@
...
+			);
+		case "call":
+			if (expression.callee === "process.exit") {
+				return `crustRuntime.Exit(${emitExpression(expression.arguments[0]!)})`;
+			}
+			return `${expression.callee}(${expression.arguments.map(emitExpression).join(", ")})`;
+		case "argv":
+			return "crustRuntime.Argv()";
+		case "slice":
+			return `${emitExpression(expression.value)}[int(${emitExpression(expression.start)}):]`;
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/emitter.ts]: [pi-watch] fixed in fb401b6

File: packages/compiler/src/frontend.ts [hunk 6]
```diff
@@ -33,30 +33,241 @@ export function lower(entryFile: string): Program {
...
+		if (ts.isFunctionDeclaration(statement)) {
+			functions.push(lowerFunction(statement, checker, sourceFile));
+		} else {
+			statements.push(lowerStatement(statement, checker, sourceFile));
+		}
 	}
-	const call = statement.expression;
+
+	return { functions, statements };
+}
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts:49]: [pi-watch] rejected: compiler package is private at 0.0.0 and is not release-versioned by Changesets

File: packages/compiler/runtime/runtime.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+)
+
+// Argv returns a Node-shaped argv while preserving user arguments from os.Args.
+func Argv() []string {
+	return append([]string{os.Args[0], os.Args[0]}, os.Args[1:]...)
+}
+
+func Index(values []string, index float64) string {
+	if index < 0 || index >= float64(len(values)) || index != math.Trunc(index) {
+		return "undefined"
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in 13913866 — unsafe length access on indexed argv values is now rejected during lowering rather than treating undefined as a string.

File: packages/compiler/runtime/runtime.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+func String(value any) string {
+	switch value := value.(type) {
+	case string:
+		return value
+	case bool:
+		return strconv.FormatBool(value)
+	case float64:
+		return strconv.FormatFloat(value, 'g', -1, 64)
+	default:
+		panic(fmt.Sprintf("unsupported JavaScript string conversion for %T", value))
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go:155]: [pi-watch] fixed in 13913866 — string-array interpolation now uses JavaScript comma-joined coercion, covered by the differential corpus.

File: packages/compiler/runtime/runtime.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+	if start < 0 {
+		start = math.Max(float64(len(values))+start, 0)
+	} else {
+		start = math.Min(start, float64(len(values)))
+	}
+	return values[int(start):]
+}
+
+func Exit(code float64) {
+	os.Exit(int(code))
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go:62]: [pi-watch] fixed in 13913866 — fractional exit codes now produce a RangeError and status 1; the corpus covers the reported case.

File: packages/compiler/runtime/runtime.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+func Log(values ...any) {
+	fmt.Println(values...)
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in 13913866 — directly logged numbers now use ECMAScript spellings, including Infinity, NaN, and negative zero.

File: packages/compiler/src/emitter.ts [hunk 11]
```diff
@@ -1,16 +1,108 @@
...
+			if (expression.callee === "process.exit") {
+				return `crustRuntime.Exit(${emitExpression(expression.arguments[0]!)})`;
+			}
+			return `${goIdentifier(expression.callee)}(${expression.arguments.map(emitExpression).join(", ")})`;
+		case "argv":
+			return "crustRuntime.Argv()";
+		case "slice":
+			return `crustRuntime.Slice(${emitExpression(expression.value)}, ${emitExpression(expression.start)})`;
+		case "length":
+			return `float64(len(${emitExpression(expression.value)}))`;
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/emitter.ts]: [pi-watch] fixed in 13913866 — emitted length operations now count UTF-16 code units for strings and elements for arrays.

File: packages/compiler/src/frontend.ts [hunk 12]
```diff
@@ -33,30 +33,245 @@ export function lower(entryFile: string): Program {
...
+				literal: span.literal.text,
+			})),
+		};
+	}
+	if (isProcessArgv(node)) return { kind: "argv" };
+	if (ts.isPropertyAccessExpression(node) && node.name.text === "length") {
+		const receiver = ts.isNonNullExpression(node.expression)
+			? node.expression.expression
+			: node.expression;
+		if (ts.isElementAccessExpression(receiver)) throw unsupported(node, sourceFile);
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 0f5af3d7 — valid indexed string lengths now use a dedicated runtime helper; undefined indexes still raise TypeError instead of becoming the string undefined.

File: packages/compiler/runtime/runtime.go [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+	if start < 0 {
+		start = math.Max(float64(len(values))+start, 0)
+	} else {
+		start = math.Min(start, float64(len(values)))
+	}
+	return values[int(start):]
+}
+
+func Exit(code float64) {
+	if code != math.Trunc(code) {
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in a4446812 — process.exit now rejects NaN and both infinities before integer conversion, with regression coverage.

File: packages/compiler/src/emitter.ts [hunk 14]
```diff
@@ -1,20 +1,112 @@
...
+			const left = emitExpression(expression.left);
+			const right = emitExpression(expression.right);
+			if (expression.type === "string") {
+				return `(crustRuntime.String(${left}) + crustRuntime.String(${right}))`;
+			}
+			if (expression.operator === "%") return `crustRuntime.Mod(${left}, ${right})`;
+			return `(${left} ${expression.operator} ${right})`;
+		}
+		case "unary":
+			return `(${expression.operator}${emitExpression(expression.operand)})`;
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/emitter.ts:65]: [pi-watch] fixed in a4446812 — unary minus now applies to a runtime float64 value, preserving the sign bit for negative zero.

File: packages/compiler/runtime/runtime.go [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+			fmt.Print(" ")
+		}
+		if number, ok := value.(float64); ok {
+			if number == 0 && math.Signbit(number) {
+				fmt.Print("-0")
+			} else {
+				fmt.Print(numberString(number))
+			}
+		} else {
+			fmt.Print(value)
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in a4446812 — direct array-valued console.log arguments are now rejected during lowering; Node-style array inspection remains outside this branch’s supported surface.

File: packages/compiler/src/emitter.ts [hunk 16]
```diff
@@ -1,20 +1,112 @@
...
+		case "identifier":
+			return goIdentifier(expression.name);
+		case "binary": {
+			const left = emitExpression(expression.left);
+			const right = emitExpression(expression.right);
+			if (expression.type === "string") {
+				return `(crustRuntime.String(${left}) + crustRuntime.String(${right}))`;
+			}
+			if (expression.operator === "%") return `crustRuntime.Mod(${left}, ${right})`;
+			return `(${left} ${expression.operator} ${right})`;
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/emitter.ts:62]: [pi-watch] fixed in a4446812 — numeric literals now cross a runtime float64 boundary before arithmetic, covering 0.1 + 0.2 and runtime overflow/division behavior.

File: packages/compiler/src/frontend.ts [hunk 17]
```diff
@@ -33,30 +33,251 @@ export function lower(entryFile: string): Program {
...
+				expression: lowerExpression(span.expression, checker, sourceFile),
+				literal: span.literal.text,
+			})),
+		};
+	}
+	if (isProcessArgv(node)) return { kind: "argv" };
+	if (ts.isPropertyAccessExpression(node) && node.name.text === "length") {
+		const receiver = ts.isNonNullExpression(node.expression)
+			? node.expression.expression
+			: node.expression;
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 087ebeaf — length receivers now unwrap parentheses and non-null assertions.

File: packages/compiler/runtime/runtime.go [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+)
+
+// Argv returns a Node-shaped argv while preserving user arguments from os.Args.
+func Argv() []string {
+	return append([]string{os.Args[0], os.Args[0]}, os.Args[1:]...)
+}
+
+func Index(values []string, index float64) string {
+	if index < 0 || index >= float64(len(values)) || index != math.Trunc(index) {
+		return "undefined"
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in 087ebeaf — out-of-range indexed values remain tagged through string parameters and throw on length access.

File: packages/compiler/runtime/runtime.go [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+	if start < 0 {
+		start = math.Max(float64(len(values))+start, 0)
+	} else {
+		start = math.Min(start, float64(len(values)))
+	}
+	return values[int(start):]
+}
+
+func Exit(code float64) {
+	if math.IsNaN(code) || math.IsInf(code, 0) || code != math.Trunc(code) {
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in 087ebeaf — process.exit now rejects values outside the JavaScript safe-integer range.

File: packages/compiler/src/frontend.ts [hunk 20]
```diff
@@ -33,30 +33,251 @@ export function lower(entryFile: string): Program {
...
-	if (!argument || !ts.isStringLiteral(argument)) {
-		throw new Error("The tracer compiler only supports a string literal console.log argument");
+	const signature = checker.getSignatureFromDeclaration(node);
+	if (!signature) throw unsupported(node, sourceFile);
+
+	return {
+		name: node.name.text,
+		parameters: node.parameters.map((parameter) => {
+			if (!ts.isIdentifier(parameter.name) || parameter.dotDotDotToken || parameter.questionToken) {
+				throw unsupported(parameter, sourceFile);
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 087ebeaf — default parameters are rejected during lowering before Go emission.

File: packages/compiler/src/frontend.ts [hunk 21]
```diff
@@ -33,30 +33,257 @@ export function lower(entryFile: string): Program {
...
+			kind: "index",
+			value: lowerExpression(node.expression, checker, sourceFile),
+			index: lowerExpression(node.argumentExpression, checker, sourceFile),
+		};
+	}
+	if (ts.isCallExpression(node)) {
+		if (
+			ts.isPropertyAccessExpression(node.expression) &&
+			node.expression.name.text === "slice" &&
+			node.arguments.length === 1
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 0976dbe3: slice lowering now requires a string-array receiver.

File: packages/compiler/src/frontend.ts [hunk 22]
```diff
@@ -33,30 +33,257 @@ export function lower(entryFile: string): Program {
...
+		return {
+			kind: "length",
+			value: lowerExpression(receiver, checker, sourceFile),
+		};
+	}
+	if (ts.isElementAccessExpression(node) && node.argumentExpression) {
+		return {
+			kind: "index",
+			value: lowerExpression(node.expression, checker, sourceFile),
+			index: lowerExpression(node.argumentExpression, checker, sourceFile),
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 0976dbe3: element access now requires a supported string-array receiver.

File: packages/compiler/src/frontend.ts [hunk 23]
```diff
@@ -33,30 +33,257 @@ export function lower(entryFile: string): Program {
...
+		return lowerExpression(node.expression, checker, sourceFile);
+	}
+	if (ts.isPrefixUnaryExpression(node)) {
+		const operator =
+			node.operator === ts.SyntaxKind.PlusToken
+				? "+"
+				: node.operator === ts.SyntaxKind.MinusToken
+					? "-"
+					: undefined;
+		if (!operator) throw unsupported(node, sourceFile);
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 0976dbe3: unary numeric operators now reject nonnumeric operands during lowering.

File: packages/compiler/src/frontend.ts [hunk 24]
```diff
@@ -33,30 +33,257 @@ export function lower(entryFile: string): Program {
...
+				kind: "call",
+				callee: "process.exit",
+				arguments: [lowerExpression(argument, checker, sourceFile)],
+			};
+		}
+		if (ts.isIdentifier(node.expression)) {
+			return {
+				kind: "call",
+				callee: node.expression.text,
+				arguments: node.arguments.map((argument) => lowerExpression(argument, checker, sourceFile)),
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 0976dbe3: identifier calls now require a source function declaration.

File: packages/compiler/src/frontend.ts [hunk 25]
```diff
@@ -33,30 +33,257 @@ export function lower(entryFile: string): Program {
...
+	}
+	if (isProcessArgv(node)) return { kind: "argv" };
+	if (ts.isPropertyAccessExpression(node) && node.name.text === "length") {
+		let receiver: ts.Expression = node.expression;
+		while (ts.isParenthesizedExpression(receiver) || ts.isNonNullExpression(receiver)) {
+			receiver = receiver.expression;
+		}
+		return {
+			kind: "length",
+			value: lowerExpression(receiver, checker, sourceFile),
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 0976dbe3: length lowering now accepts only supported string and string-array receivers.

File: packages/compiler/runtime/runtime.go [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+	"math"
+	"os"
+	"strconv"
+	"strings"
+	"unicode/utf16"
+)
+
+// Argv returns a Node-shaped argv while preserving user arguments from os.Args.
+func Argv() []string {
+	return append([]string{os.Args[0], os.Args[0]}, os.Args[1:]...)
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in 0976dbe3: generated argv now preserves the TypeScript entry path at index 1.

File: packages/compiler/src/emitter.ts [hunk 27]
```diff
@@ -1,20 +1,118 @@
...
+			}
+			return `${goIdentifier(expression.callee)}(${expression.arguments.map(emitExpression).join(", ")})`;
+		case "argv":
+			return "crustRuntime.Argv()";
+		case "slice":
+			return `crustRuntime.Slice(${emitExpression(expression.value)}, ${emitExpression(expression.start)})`;
+		case "length":
+			if (expression.value.kind === "index") {
+				return `crustRuntime.IndexLength(${emitExpression(expression.value.value)}, ${emitExpression(expression.value.index)})`;
+			}
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/emitter.ts:84]: [pi-watch] fixed in 0976dbe3: unsupported string element access is rejected before emission, including indexed length.

File: packages/compiler/runtime/runtime.go [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+	default:
+		panic(fmt.Sprintf("unsupported JavaScript length for %T", value))
+	}
+}
+
+func Log(values ...any) {
+	for index, value := range values {
+		if index > 0 {
+			fmt.Print(" ")
+		}
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go:84]: [pi-watch] fixed in 3bcb2d03: console.log calls with format tokens are now rejected during lowering.

File: packages/compiler/src/frontend.ts [hunk 29]
```diff
@@ -33,30 +33,284 @@ export function lower(entryFile: string): Program {
...
+		throw unsupported(node, sourceFile);
+	}
+
+	const call = node.expression;
+	if (isPropertyCall(call, "console", "log")) {
+		if (
+			call.arguments.length === 0 ||
+			call.arguments.some((argument) =>
+				checkerTypeIsStringArray(checker.getTypeAtLocation(argument)),
+			)
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 3bcb2d03: function-valued identifiers are now rejected during lowering.

File: packages/compiler/src/frontend.ts [hunk 30]
```diff
@@ -33,30 +33,294 @@ export function lower(entryFile: string): Program {
...
+		return { kind: "literal", type: "number", value: Number(node.text) };
+	}
+	if (node.kind === ts.SyntaxKind.TrueKeyword || node.kind === ts.SyntaxKind.FalseKeyword) {
+		return { kind: "literal", type: "boolean", value: node.kind === ts.SyntaxKind.TrueKeyword };
 	}
+	if (ts.isIdentifier(node)) {
+		if (checker.getTypeAtLocation(node).getCallSignatures().length > 0) {
+			throw unsupported(node, sourceFile);
+		}
+		return { kind: "identifier", name: node.text };
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts:188]: [pi-watch] fixed in 16fd6deb — identifier lowering now accepts only source parameters, rejecting unsupported globals before emission.

File: packages/compiler/src/frontend.ts [hunk 31]
```diff
@@ -33,30 +33,294 @@ export function lower(entryFile: string): Program {
...
+		const firstArgument = call.arguments[0];
+		if (
+			call.arguments.length === 0 ||
+			(call.arguments.length > 1 &&
+				firstArgument &&
+				ts.isStringLiteralLike(firstArgument) &&
+				/%[cdfijoOs%]/.test(firstArgument.text)) ||
+			call.arguments.some((argument) =>
+				checkerTypeIsStringArray(checker.getTypeAtLocation(argument)),
+			)
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 16fd6deb — void/undefined console arguments are rejected during lowering.

File: packages/compiler/src/frontend.ts [hunk 32]
```diff
@@ -33,30 +33,294 @@ export function lower(entryFile: string): Program {
...
+		};
+	}
+	if (ts.isElementAccessExpression(node) && node.argumentExpression) {
+		if (!checkerTypeIsStringArray(checker.getTypeAtLocation(node.expression))) {
+			throw unsupported(node, sourceFile);
+		}
+		return {
+			kind: "index",
+			value: lowerExpression(node.expression, checker, sourceFile),
+			index: lowerExpression(node.argumentExpression, checker, sourceFile),
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 16fd6deb — array indexes must now have a numeric TypeScript type.

File: packages/compiler/src/frontend.ts [hunk 33]
```diff
@@ -33,30 +33,294 @@ export function lower(entryFile: string): Program {
...
+				start: lowerExpression(start, checker, sourceFile),
+			};
+		}
+		if (isPropertyCall(node, "process", "exit") && node.arguments.length === 1) {
+			const argument = node.arguments[0];
+			if (!argument) throw unsupported(node, sourceFile);
+			return {
+				kind: "call",
+				callee: "process.exit",
+				arguments: [lowerExpression(argument, checker, sourceFile)],
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 16fd6deb — process.exit arguments must now have a numeric TypeScript type.

File: packages/compiler/src/frontend.ts [hunk 34]
```diff
@@ -33,30 +33,311 @@ export function lower(entryFile: string): Program {
...
+			right: lowerExpression(node.right, checker, sourceFile),
+			type: lowerType(checker.getTypeAtLocation(node), node, sourceFile),
+		};
+	}
+	if (ts.isTemplateExpression(node)) {
+		return {
+			kind: "template",
+			head: node.head.text,
+			spans: node.templateSpans.map((span) => ({
+				expression: lowerExpression(span.expression, checker, sourceFile),
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 301e61e4: void/undefined template interpolations are rejected before emission, with regression coverage.

File: packages/compiler/src/frontend.ts [hunk 35]
```diff
@@ -33,30 +33,311 @@ export function lower(entryFile: string): Program {
...
+
+	const call = node.expression;
+	if (isPropertyCall(call, "console", "log")) {
+		const firstArgument = call.arguments[0];
+		if (
+			call.arguments.length === 0 ||
+			(call.arguments.length > 1 &&
+				firstArgument &&
+				ts.isStringLiteralLike(firstArgument) &&
+				/%[cdfijoOs%]/.test(firstArgument.text)) ||
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 301e61e4: multi-argument logs now reject nonliteral string format candidates; the corpus keeps a literal first argument.

File: packages/compiler/src/emitter.ts [hunk 36]
```diff
@@ -1,20 +1,118 @@
...
+			if (expression.type === "string") return goString(expression.value as string);
+			if (expression.type === "boolean") return String(expression.value);
+			return emitNumber(expression.value as number);
+		case "identifier":
+			return goIdentifier(expression.name);
+		case "binary": {
+			const left = emitExpression(expression.left);
+			const right = emitExpression(expression.right);
+			if (expression.type === "string") {
+				return `(crustRuntime.String(${left}) + crustRuntime.String(${right}))`;
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/emitter.ts]: [pi-watch] fixed in 6cf2efec: string-typed additions now choose concatenation versus numeric addition from runtime operands.

File: packages/compiler/src/frontend.ts [hunk 37]
```diff
@@ -33,30 +33,323 @@ export function lower(entryFile: string): Program {
...
+					? /%[cdfijoOs%]/.test(firstArgument.text)
+					: Boolean(
+							checker.getTypeAtLocation(firstArgument).flags &
+								(ts.TypeFlags.String | ts.TypeFlags.StringLiteral),
+						))) ||
+			call.arguments.some((argument) => {
+				const type = checker.getTypeAtLocation(argument);
+				return (
+					checkerTypeIsStringArray(type) ||
+					Boolean(type.flags & (ts.TypeFlags.Void | ts.TypeFlags.Undefined))
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 6cf2efec: never-valued console arguments are rejected before emission.

File: packages/compiler/src/frontend.ts [hunk 38]
```diff
@@ -33,30 +33,323 @@ export function lower(entryFile: string): Program {
...
+				parameter.dotDotDotToken ||
+				parameter.questionToken ||
+				parameter.initializer
+			) {
+				throw unsupported(parameter, sourceFile);
+			}
+			return {
+				name: parameter.name.text,
+				type: lowerType(checker.getTypeAtLocation(parameter), parameter, sourceFile),
+			};
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts:90]: [pi-watch] fixed in 6cf2efec: void/undefined parameters are rejected before emitting Go signatures.

File: packages/compiler/src/frontend.ts [hunk 39]
```diff
@@ -33,30 +33,323 @@ export function lower(entryFile: string): Program {
...
+function lowerStatement(
+	node: ts.Statement,
+	checker: ts.TypeChecker,
+	sourceFile: ts.SourceFile,
+): Statement {
+	if (ts.isReturnStatement(node)) {
+		return {
+			kind: "return",
+			expression: node.expression
+				? lowerExpression(node.expression, checker, sourceFile)
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/src/frontend.ts]: [pi-watch] fixed in 6cf2efec: void/never-valued return expressions are rejected before emission.

File: packages/compiler/runtime/runtime.go [hunk 40]
```diff
@@ -0,0 +1,156 @@
...
+	"math"
+	"os"
+	"strconv"
+	"strings"
+	"unicode/utf16"
+)
+
+// Argv returns a Node-shaped argv while preserving user arguments from os.Args.
+func Argv(entryFile string) []string {
+	return append([]string{os.Args[0], entryFile}, os.Args[1:]...)
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in 6cf2efec: argv[0] now uses the canonical running executable path.
